geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/collections/OrderStatisticsTree.java

```diff
@@ -0,0 +1,809 @@
+          root = subTree;
+        } else if (grandParent.left == parent) {
+          grandParent.left = subTree;
+        } else {
+          grandParent.right = subTree;
+        }
+
+        if (grandParent != null) {
+          grandParent.height =
+              Math.max(height(grandParent.left),
+                  height(grandParent.right)) + 1;
+        }
+
+        if (insertionMode) {
+          return;
+        }
+      }
+
+      parent.height = Math.max(height(parent.left),
+          height(parent.right)) + 1;
+      parent = parent.parent;
+    }
+  }
+
+  public boolean isHealthy() {
+    if (root == null) {
+      return true;
+    }
+
+    return !containsCycles()
+        && heightsAreCorrect()
+        && isBalanced()
+        && isWellIndexed();
+  }
+
+  private boolean containsCycles() {
+    Set<Node<T>> visitedNodes = new HashSet<>();
+    return containsCycles(root, visitedNodes);
+  }
+
+  private boolean containsCycles(Node<T> current, Set<Node<T>> visitedNodes) {
+    if (current == null) {
+      return false;
+    }
+
+    if (visitedNodes.contains(current)) {
+      return true;
+    }
+
+    visitedNodes.add(current);
+
+    return containsCycles(current.left, visitedNodes)
+        || containsCycles(current.right, visitedNodes);
+  }
+
+  private boolean heightsAreCorrect() {
+    return getHeight(root) == root.height;
+  }
+
+  private int getHeight(Node<T> node) {
+    if (node == null) {
+      return -1;
+    }
+
+    int leftTreeHeight = getHeight(node.left);
+
+    if (leftTreeHeight == Integer.MIN_VALUE) {
+      return Integer.MIN_VALUE;
+    }
+
+    int rightTreeHeight = getHeight(node.right);
+
+    if (rightTreeHeight == Integer.MIN_VALUE) {
+      return Integer.MIN_VALUE;
+    }
+
+    if (node.height == Math.max(leftTreeHeight, rightTreeHeight) + 1) {
+      return node.height;
+    }
+
+    return Integer.MIN_VALUE;
+  }
+
+  private boolean isBalanced() {
+    return isBalanced(root);
+  }
+
+  private boolean isBalanced(Node<T> node) {
+    if (node == null) {
+      return true;
+    }
+
+    if (!isBalanced(node.left)) {
+      return false;
+    }
+
+    if (!isBalanced(node.right)) {
+      return false;
+    }
+
+    int leftHeight = height(node.left);
+    int rightHeight = height(node.right);
+
+    return Math.abs(leftHeight - rightHeight) < 2;
+  }
+
+  private boolean isWellIndexed() {
+    return size == count(root);
+  }
+
+  private int count(Node<T> node) {
+    if (node == null) {
+      return 0;
+    }
+
+    int leftTreeSize = count(node.left);
+
+    if (leftTreeSize == Integer.MIN_VALUE) {
+      return Integer.MIN_VALUE;
+    }
+
+    if (node.count != leftTreeSize) {
+      return Integer.MIN_VALUE;
+    }
+
+    int rightTreeSize = count(node.right);
+
+    if (rightTreeSize == Integer.MIN_VALUE) {
+      return Integer.MIN_VALUE;
+    }
+
+    return leftTreeSize + 1 + rightTreeSize;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
I mentioned this earlier, but leaving this here as well. hashCode() and equals() for this class should obey the set contract. See the javadocs for Set.equals() and Set.hashCode()

Reply:
They're tweaked, and pass all existing tests on my machine. We'll see what LGTM has to say.